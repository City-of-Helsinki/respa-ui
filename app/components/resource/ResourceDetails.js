import _ from 'lodash';
import React, { Component, PropTypes } from 'react';

export class ResourceDetails extends Component {
  render() {
    const { capacityString, description, type } = this.props;

    return (
      <div>
        <p>{_.capitalize(type)} {capacityString}</p>
        <p>{description}</p>
      </div>
    );
  }
}

ResourceDetails.propTypes = {
  capacityString: PropTypes.string.isRequired,
  description: PropTypes.string.isRequired,
  type: PropTypes.string.isRequired,
};

export default ResourceDetails;
