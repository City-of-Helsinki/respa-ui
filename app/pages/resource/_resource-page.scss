$carousel-image-min-height: 260px;
$carousel-image-mid-height: 350px;
$carousel-image-max-height: 420px;

@import './reservation-calendar/reservation-calendar';
@import './reservation-calendar/time-slots/time-slots';
@import './resource-info/image-carousel';
@import './resource-info/resource-info';

.app-ResourcePage {
  max-width: 700px;

  .date-header {
    margin-bottom: 20px;
    font-size: 1.8rem;
    color: $brand-primary;
    text-align: left;
  }
}
