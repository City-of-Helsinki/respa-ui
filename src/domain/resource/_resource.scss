@import './card/resourceCard';
@import './card/slider/resourceCardSlider';
