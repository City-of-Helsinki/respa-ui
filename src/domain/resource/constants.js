export const resourceProductTypes = Object.freeze({
  RENT: 'rent',
  EXTRA: 'extra',
});

export const resourcePriceTypes = Object.freeze({
  HOURLY: 'hourly',
  DAILY: 'daily',
  WEEKLY: 'weekly',
  FIXED: 'fixed',
});

export const productPriceType = Object.freeze({
  PER_PERIOD: 'per_period',
  FIXED: 'fixed',
});

export const resourcePeriodLengthType = Object.freeze({
  WITHIN_DAY: 'within_day',
  WHOLE_DAY: 'whole_day',
  OVERNIGHT: 'over_night',
});
