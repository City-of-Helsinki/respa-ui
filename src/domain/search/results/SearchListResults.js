import React from 'react';
import PropTypes from 'prop-types';

class SearchListResults extends React.Component {
  static propTypes = {
    resources: PropTypes.array,
    isLoading: PropTypes.bool,
  };

  render() {
    const { isLoading, resources } = this.props;
    return (
      <div className="app-SearchListResults">
        SearchListResults
        <ul>
          {resources && resources.map((item, i) => <li key={`item-${i}`}>{item.id}: {item.name.fi}</li>)}
        </ul>
      </div>
    );
  }
}

export default SearchListResults;
