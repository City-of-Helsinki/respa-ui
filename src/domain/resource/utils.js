import round from 'lodash/round';
import filter from 'lodash/filter';
import findIndex from 'lodash/findIndex';
import get from 'lodash/get';
import find from 'lodash/find';
import moment from 'moment';

import * as urlUtils from '../../common/url/utils';
import * as dataUtils from '../../common/data/utils';
import * as reservationUtils from '../reservation/utils';
import constants from '../../../app/constants/AppConstants';
import forEach from "lodash/forEach";


/**
 * Getter for resource page link.
 * Returns a resource page url with the given query params.
 * @param resource
 * @param query
 * @returns {string}
 */
export const getResourcePageLink = (resource, query) => {
  return urlUtils.getLinkString(`/resources/${resource.id}`, query);
};

/**
 * Getter for formatted unit address string.
 * @param unit
 * @param locale
 * @returns {string}
 */
export const getUnitAddress = (unit, locale) => {
  let postal;
  const streetAddress = dataUtils.getLocalizedFieldValue(unit.street_address, locale);
  if (unit && unit.address_zip) {
    postal = unit.address_zip;
  }
  if (unit && unit.municipality) {
    postal = postal ? `${postal} ${unit.municipality}` : unit.municipality;
  }

  return `${streetAddress}${postal ? ` ${postal}` : ''}`;
};

/**
 * Getter for formatted resource distance string.
 * @param resource
 * @returns {string}
 */
export const getResourceDistance = (resource) => {
  const km = resource.distance / 1000;
  return km ? `${round(km, km < 10 ? 1 : null)} km` : '';
};

/**
 * Getter for price string used in resource cards.
 * @param resource
 * @param t
 * @returns {string|*}
 */
export const getPrice = (resource, t) => {
  const minPricePerHour = resource.min_price_per_hour;
  const maxPricePerHour = resource.max_price_per_hour;

  if (!(minPricePerHour || maxPricePerHour)) {
    return t('ResourceIcons.free');
  }

  if (minPricePerHour && maxPricePerHour && minPricePerHour !== maxPricePerHour) {
    return `${Number(minPricePerHour)} - ${Number(maxPricePerHour)} €/h`;
  }

  const priceString = maxPricePerHour || minPricePerHour;
  const price = priceString !== 0 ? Number(priceString) : 0;
  if (price === 0) {
    return t('ResourceIcons.free');
  }

  return price ? `${price} €/h` : null;
};

export const isFree = (resource) => {
  const minPricePerHour = resource.min_price_per_hour;
  const maxPricePerHour = resource.max_price_per_hour;

  if (!(minPricePerHour || maxPricePerHour)) {
    return true;
  }

  if (minPricePerHour && maxPricePerHour && minPricePerHour !== maxPricePerHour) {
    return false;
  }

  const priceString = maxPricePerHour || minPricePerHour;
  const price = priceString !== 0 ? Number(priceString) : 0;

  return price === 0;
};

export const getOpeningHours = (resource, date = null) => {
  if (!resource) {
    return null;
  }

  const openingHours = get(resource, 'opening_hours', []);
  if (!openingHours.length) {
    return null;
  }

  const index = date ? findIndex(openingHours, item => item.date === date) : 0;
  return {
    closes: get(openingHours, `[${index}].closes`),
    opens: get(openingHours, `[${index}].opens`),
  };
};

const reservingIsRestricted = (resource, date) => {
  if (!date) {
    return false;
  }

  const isAdmin = get(resource, 'user_permissions.is_admin', false);
  const isLimited = resource.reservable_before && moment(resource.reservable_before).isBefore(moment(date), 'day');

  return isLimited && !isAdmin;
};

export const getOpenReservations = (resource) => {
  return filter(resource.reservations, (reservation) => {
    return reservation.state !== 'cancelled' && reservation.state !== 'denied';
  });
};

export const getAvailabilityDataForNow = (resource, date) => {
  const openingHours = getOpeningHours(resource, date);
  const reservations = getOpenReservations(resource);

  if (!openingHours || !openingHours.closes || !openingHours.opens) {
    return { status: 'closed', bsStyle: 'danger' };
  }

  const nowMoment = moment();
  const opensMoment = moment(openingHours.opens);
  const closesMoment = moment(openingHours.closes);
  const beginMoment = nowMoment > opensMoment ? nowMoment : opensMoment;

  if (nowMoment > closesMoment) {
    return { status: 'closed', bsStyle: 'danger' };
  }

  const currentReservation = reservationUtils.getCurrentReservation(reservations);
  if (currentReservation || nowMoment < opensMoment) {
    const nextAvailableTime = reservationUtils.getNextAvailableTime(reservations, beginMoment);
    if (nextAvailableTime < closesMoment) {
      return {
        status: 'availableAt',
        bsStyle: 'danger',
        values: { time: nextAvailableTime.format(constants.TIME_FORMAT) },
      };
    }
    return { status: 'reserved', bsStyle: 'danger' };
  }

  return { status: 'available', bsStyle: 'success' };
};

export const getAvailabilityDataForWholeDay = (resource, date) => {
  const openingHours = getOpeningHours(resource, date);

  if (!openingHours || !openingHours.closes || !openingHours.opens) {
    return { status: 'closed', bsStyle: 'danger' };
  }

  if (reservingIsRestricted(resource, date)) {
    return { status: 'reservingRestricted', bsStyle: 'danger' };
  }

  const reservations = getOpenReservations(resource);

  const opensMoment = moment(openingHours.opens);
  const closesMoment = moment(openingHours.closes);
  let total = closesMoment - opensMoment;

  forEach(reservations, (reservation) => {
    const resBeginMoment = moment(reservation.begin);
    if (!resBeginMoment.isSame(opensMoment, 'd')) {
      return;
    }

    const resEndMoment = moment(reservation.end);
    total -= resEndMoment;
    total += resBeginMoment;
  });

  const asHours = moment.duration(total).asHours();
  const rounded = Math.ceil(asHours * 2) / 2;

  if (rounded === 0) {
    return { status: 'reserved', bsStyle: 'danger' };
  }

  return {
    status: 'availableTime',
    bsStyle: 'success',
    values: { hours: rounded },
  };
};
