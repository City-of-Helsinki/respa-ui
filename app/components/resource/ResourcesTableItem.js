import React, { Component, PropTypes } from 'react';
import moment from 'moment';
import { Link } from 'react-router';

import TimeRange from 'components/common/TimeRange';
import { getOpeningHours } from 'utils/DataUtils';

class ResourcesTableItem extends Component {
  getReserverName(reservation) {
    return (
      reservation.reserverName ||
      (reservation.user &&
        ((reservation.user.displayName !== '' && reservation.user.displayName) ||
        reservation.user.email)
      )
    );
  }

  getAvailableTime(untilDate) {
    const availableTime = moment(untilDate).diff(moment(), 'hours', true).toFixed(1);
    return (
      <td className="resource-table-row available">{
        `${availableTime}h heti`
      }</td>
    );
  }

  renderAvailable() {
    const {
      currentReservation,
      nextReservation,
      resource,
    } = this.props;
    if (currentReservation) {
      return <td className="resource-table-row available reserved">Varattu</td>;
    }
    if (nextReservation) {
      return this.getAvailableTime(nextReservation.begin);
    }
    const closes = getOpeningHours(resource).closes;
    if (moment() < moment(closes)) {
      return this.getAvailableTime(closes);
    }
    return <td className="resource-table-row available" />;
  }

  render() {
    const {
      currentReservation,
      nextReservation,
      resource,
    } = this.props;
    const reservation = currentReservation || nextReservation;
    return (
      <tr>
        <td className="resource-table-row name">
          <Link to={`/resources/${resource.id}`}>
            {resource.name.fi}
          </Link>
        </td>
        {this.renderAvailable()}
        { reservation ? [
          <td className="resource-table-row reservation-range" key={`${reservation.id}-range`}>
            <TimeRange
              begin={reservation.begin}
              dateFormat=" "
              end={reservation.end}
            />
          </td>,
          <td className="resource-table-row reserver" key={`${reservation.id}-reserver`}>
            {this.getReserverName(reservation)}
          </td>,
          <td className="resource-table-row comments" key={`${reservation.id}-comments`}>
            {reservation.comments}
          </td>,
        ] : [
          <td className="resource-table-row reservation-range" key="no-res-range" />,
          <td className="resource-table-row reserver" key="no-res-reserver" />,
          <td className="resource-table-row comments" key="no-res-comments" />,
        ]}
      </tr>
    );
  }
}

const reservationPropType = PropTypes.shape({
  begin: PropTypes.string.isRequired,
  end: PropTypes.string.isRequired,
  reserverName: PropTypes.string,
  comments: PropTypes.arrayOf(PropTypes.string),
});

ResourcesTableItem.propTypes = {
  currentReservation: reservationPropType,
  nextReservation: reservationPropType,
  resource: PropTypes.shape({
    openingHours: PropTypes.array.isRequired,
    user: PropTypes.shape({
      displayName: PropTypes.string,
      email: PropTypes.string.isRequired,
    }),
  }).isRequired,
};

export default ResourcesTableItem;
