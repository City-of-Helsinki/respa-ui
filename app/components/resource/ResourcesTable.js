import React, { Component, PropTypes } from 'react';
import Table from 'react-bootstrap/lib/Table';
import { connect } from 'react-redux';

import ResourcesTableItem from 'components/resource/ResourcesTableItem';
import { getCurrentReservation, getNextReservation } from 'utils/reservationUtils';

export class UnconnectedResourcesTable extends Component {
  renderResourcesTableItem(resource) {
    return (
      <ResourcesTableItem
        key={resource.id}
        currentReservation={getCurrentReservation(resource.reservations)}
        nextReservation={getNextReservation(resource.reservations)}
        resource={resource}
      />
    );
  }

  render() {
    const {
      emptyMessage,
      resources,
    } = this.props;
    return (
      Object.keys(resources).length ? (
        <Table className="resources-table" responsive striped>
          <thead>
            <tr>
              <th>Tilan nimi</th>
              <th>Vapaata</th>
              <th>Varaus</th>
              <th>Varaaja</th>
              <th>Kommentit</th>
            </tr>
          </thead>
          <tbody>
            {resources.map(this.renderResourcesTableItem)}
          </tbody>
        </Table>
      ) : (
        <p>{emptyMessage || 'Et ole lis채nnyt viel채 yht채채n tilaa itsellesi.'}</p>
      )
    );
  }
}

UnconnectedResourcesTable.propTypes = {
  emptyMessage: PropTypes.string,
  resources: PropTypes.array.isRequired,
};

export default connect()(UnconnectedResourcesTable);
