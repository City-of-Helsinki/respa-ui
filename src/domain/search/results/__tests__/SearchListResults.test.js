import React from 'react';
import toJSON from 'enzyme-to-json';
import { shallow } from 'enzyme';

import SearchListResults from '../SearchListResults';
import resource from '../../../../common/data/fixtures/resource';
import unit from '../../../../common/data/fixtures/unit';

describe('SearchListResults', () => {
  test('renders correctly', () => {
    const wrapper = shallow(
      <SearchListResults
        onFavoriteClick={() => null}
        onFiltersChange={() => null}
        resources={[resource.build()]}
        units={[unit.build()]}
      />
    );

    expect(toJSON(wrapper)).toMatchSnapshot();
  });
});
