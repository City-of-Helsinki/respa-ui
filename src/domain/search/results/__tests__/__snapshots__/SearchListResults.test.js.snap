// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchListResults renders correctly 1`] = `
<Route>
  <Component />
</Route>
`;
